### PR TITLE
Fix historical ORC timestamp conversion

### LINKED ISSUE #15895: [BUG] orc_test.py::test_read_round_trip fails with DATAGEN_SEED=1788463960 — GPU/CPU timestamp mismatch at the 1883 LMT/EST transition
### Describe the bug

`orc_test.py::test_read_round_trip` fails with a GPU/CPU value mismatch when the data generator produces a timestamp that falls on the 1883 Local Mean Time to standard time transition and the tests run under a non-UTC timezone.

The failure is data-dependent: it only appears when the generated data happens to contain a timestamp inside a narrow window around that transition, so it surfaces intermittently as the datagen seed changes from run to run.

Observed with `DATAGEN_SEED=1788463960` and `TZ=America/New_York`.

### Steps/Code to reproduce bug

The seed is pinned, so this is deterministic:

```bash
DATAGEN_SEED=1788463960 TZ=America/New_York ./integration_tests/run_pyspark_from_build.sh -k "test_read_round_trip and orc"
```

### Expected behavior

GPU and CPU return identical timestamp values for ORC reads, including for timestamps predating standard time zone adoption.

### Actual behavior

62 test failures, all carrying the byte-identical assertion:

```
AssertionError: GPU (-2717655100076025) and CPU (-2717655338076025) int values are different at [688, '_c8', 15]
```

`_c8` is the `Array(Timestamp)` column — index 8 of `orc_basic_gens`, generated by `orc_timestamp_gen`, which is `TimestampGen(start=datetime(1590, 1, 1, tzinfo=timezone.utc))`.

### Analysis

Decoding the two values as microseconds since the Unix epoch:

| | microseconds | timestamp |
|---|---|---|
| GPU | -2717655100076025 | 1883-11-18 15:48:19.923975 |
| CPU | -2717655338076025 | 1883-11-18 15:44:21.923975 |

The difference is exactly **238,000,000 µs = 238 seconds**.

That is precisely the offset difference between `America/New_York` Local Mean Time (UTC−04:56:02) and Eastern Standard Time (UTC−05:00:00):

```
(5*3600) - (4*3600 + 56*60 + 2) = 238 seconds
```

And **1883-11-18** is the date US standard time zones took effect, when New York switched from LMT to EST.

So the GPU and CPU disagree about which offset applies to a timestamp sitting at that transition boundary. One side is applying LMT, the other EST.

### Scope

All 62 failures come from a **single bad generated value** — every failure reports the same GPU value, the same CPU value, and the array element index 15, differing only in which row of the shared dataset it lands on (`[688, ...]`, `[58, ...]`, `[809, ...]`). One value in the generated data breaks every parameterization that reads that dataset.

Affected tests:

- `orc_test.py::test_read_round_trip` — 60 failures
- `orc_test.py::test_read_round_trip_for_multithreaded_combining` — 2 failures

The failure is independent of reader configuration. It reproduces across:

- all three ORC reader types — `PERFILE`, `COALESCING`, `MULTITHREADED`
- both `spark.sql.sources.useV1SourceList` settings (V1 `orc` and V2)
- both read paths — `read_orc_df` and `read_orc_sql`
- chunked reads on and off, with and without `limitMemoryUsage`

That breadth is consistent with the defect being in timestamp/timezone conversion rather than in any particular reader implementation.

### Why this is intermittent

`run_pyspark_from_build.sh` defaults the datagen seed to `` `date +%s` ``, so every run samples different data. `orc_timestamp_gen` spans 1590 → 9999, roughly 8400 years, while the window where GPU and CPU disagree appears to be very narrow — otherwise every run containing pre-1883 timestamps would fail, which is not what is observed.

So the practical effect is a low-probability, seed-dependent failure: most runs are green, and an occasional run lands on the bad window and fails.

Both affected tests are already marked `@tz_sensitive_test`.

### Environment

- `spark-rapids-jni` 26.10.0-SNAPSHOT (snapshot `26.10.0-20260903.061551-22`)
- `TZ=America/New_York`
- Failure observed in a routine integration-test run; not tied to any particular change under test

### PR DESCRIPTION
**Pull request by res-life**
closes https://github.com/NVIDIA/cudf-spark/issues/15895.

Depends on:
* https://github.com/NVIDIA/cudf-spark-jni/pull/5171)
Merge the JNI change and make a `cudf-spark-jni` artifact containing `GpuTimeZoneDB.convertOrcTimestampToSpark` and `GpuTimeZoneDB.convertOrcIntegerTimestampToSpark` available before merging this PR.

### Description

Use the fused JNI ORC-to-Spark timezone conversion for physical timestamps and integer-to-timestamp schema evolution.

#### Performance numbers: 50 million rows

On this single-host end-to-end benchmark, GPU/Legacy GPU mean speedups were 1.05–1.09x for the modern non-UTC cases and 0.97x for the UTC control; GPU/CPU mean speedups were 1.47–2.83x. These are observed means, not statistically established improvements.

Environment: NVIDIA L4; AMD EPYC 7R13, 16 logical CPUs (8 cores), 60 GiB RAM; Spark 3.5.6, Scala 2.12, Zulu JDK 17.0.17; `local[16]`, 16 GiB driver memory, six concurrent GPU tasks, and 256 MiB GPU batches.

Each dataset contains 50,000,000 rows in 32 Snappy ORC files, with approximately 0.99% null timestamps. Every mode reads the same files and evaluates `sum(hash(ts))`. Times below are arithmetic mean ± sample standard deviation in milliseconds, from 10 timed queries per mode/case.

LA = `America/Los_Angeles`; NY = `America/New_York`. Physical timestamp arrows denote writer → reader timezone. **GPU/Legacy GPU speedup = Legacy GPU time / GPU time; GPU/CPU speedup = CPU time / GPU time.** A value below 1 is slower.

| Case | CPU (ms) | Legacy GPU (ms) | GPU (ms) | GPU/Legacy GPU speedup | GPU/CPU speedup | Legacy GPU / GPU correctness |
|---|---:|---:|---:|---:|---:|---|
| Physical timestamp, LA → UTC (modern) | 576.1 ± 56.4 | 286.6 ± 16.1 | 294.8 ± 26.5 | 0.97x | 1.95x | match / match |
| Physical timestamp, LA → LA (modern) | 443.6 ± 26.8 | 329.7 ± 34.4 | 302.1 ± 29.4 | 1.09x | 1.47x | match / match |
| Physical timestamp, LA → NY (modern) | 752.7 ± 21.0 | 330.5 ± 43.6 | 303.3 ± 33.7 | 1.09x | 2.48x | match / match |
| BIGINT → TIMESTAMP, reader NY (modern) | 463.3 ± 44.4 | 228.1 ± 6.7 | 217.3 ± 14.0 | 1.05x | 2.13x | match / match |
| Physical timestamp, NY → NY (historical) | 511.4 ± 58.1 | 283.0 ± 15.7 | 256.3 ± 15.5 | N/A (Legacy GPU mismatch) | 2.00x | mismatch / match |
| BIGINT → TIMESTAMP, reader NY (historical) | 605.5 ± 70.8 | 210.8 ± 8.9 | 214.3 ± 12.8 | N/A (Legacy GPU mismatch) | 2.83x | mismatch / match |

<details>
<summary>Benchmark methodology and interpretation</summary>

- **A/B scope:** Legacy GPU uses plugin base `616bff688494856cd40b6c33005f2af37eb828fc`; GPU uses this PR's head `c71593d6bf8b6883e7759df8b7f72668686398cc`. Both use exactly the same JNI `82d67f0012e65f7d0e124a7c42e24a1ab411b949` and cuDF `5339497a1a17d799687cbf189fb113411fb015ca`. This isolates the old multi-step conversion versus the fused API, not old versus new JNI binaries.
- **Measurement:** Two independent JVM sessions per mode/timezone, ordered CPU → Legacy GPU → GPU and then GPU → Legacy GPU → CPU. Each case has a correctness scan, two warm-up queries, and five measured queries per session.
- **Timing boundary:** Starts before DataFrame/query construction and ends after `collect()`, including file listing/planning, ORC read/decode, timezone conversion, hash aggregation, shuffle, and result collection. Excludes data generation, JVM/Spark startup, explicit validation, and metric extraction.
- **Reader/cache settings:** MULTITHREADED ORC reader with 16 reader threads; 128 MiB maximum file partition size. AQE and RAPIDS file cache disabled; no DataFrame cache/persist. OS page cache is warm and is not dropped. OS, JVM, and Spark session timezones are set consistently.
- **Data:** Modern values span 2000–2033; physical timestamps include microsecond fractions. Historical physical timestamps span ±24 hours around `1883-11-18T15:00Z`, including New York's `17:00Z` transition. Historical integers span ±24 hours around local `1883-11-18 12:02:00`. Data-generation seed: `391183073`.
- **Correctness:** GPU matches CPU row count, non-null count, minimum/maximum timestamp, `sum(hash(ts))`, `sum(hash(id, ts))`, and `sum(xxhash64(id, ts))` in every case/session. These are aggregate checks, not an exact row-by-row proof. Legacy GPU mismatches CPU in both historical cases, so those rows do not report an equivalent-result GPU/Legacy GPU speedup.
- **Execution:** All 18 measurement processes completed successfully. All 252 warm-up/measured query plans report 50,000,000 scan output rows; GPU plans use RAPIDS ORC scans and GPU hash aggregation.
- **Variability:** For modern LA → NY, session means were 315.84/317.18 ms (Legacy GPU/GPU) in session 1 and 345.16/289.48 ms in session 2. The aggregate 1.09x therefore does not establish a stable 9% speedup.
- **Why the end-to-end difference is modest:** Legacy GPU already runs conversion on the GPU. Fusion reduces correction kernels and temporary columns, while ORC decoding, Spark scheduling, aggregation, and collection remain. The UTC control reuses the existing native conversion path. These measurements do not isolate timezone kernel time or establish its fraction of total query latency.

</details>

### Correctness fixes

- Fix historical physical timestamp round-trip mismatches, including New York's 1883 LMT-to-EST transition and pre-1901 Shanghai values, across scalar and nested array columns.
- Fix `BIGINT -> TIMESTAMP` at historical overlaps by preserving the instant selected by Apache ORC's `java.util.TimeZone` conversion through Spark's `java.time` rebase. The New York `1883-11-18 12:02:00` case now matches CPU Spark's later EST occurrence.
- Cover timezones such as `Pacific/Port_Moresby` whose legacy `TimeZone` has no transition table but whose `java.time` rules contain historical LMT offsets.
- Preserve exact negative sub-second behavior at historical gap boundaries.
- Restore the general ORC timestamp generator lower bound from 1901 to 1590 and add deterministic physical and integer historical-transition regressions.

### Performance and structure

- Replace the multi-kernel temporary-column correction in `GpuOrcTimezoneUtils` with one JNI call per timestamp leaf.
- Remove intermediate casts, arithmetic columns, comparisons, and `ifElse` results from the hot read path.
- Ratchet the resolved resource-nesting baseline after removing the old branch.

### Validation

- JNI Docker package with native test targets enabled: `BUILD SUCCESS`.
- Historical-overlap GPU audit: 129 timezones, 655 boundary values, 0 mismatches against the CPU/Spark reference.
- Spark 3.5.6 package: `BUILD SUCCESS`.
- `test_casting_from_bigint_to_timestamp_in_new_york_lmt_overlap`: passed as a normal end-to-end GPU/CPU equality test; the previous strict expected-failure marker was removed.
- Physical New York reader matrix and `OrcTimezoneSuite` historical coverage passed in the fused-path validation.
- Existing `OrcTimezonePerfSuite` 10,000,000-row measurement: CPU/GPU results matched; cross-timezone GPU mean 330 ms, same-timezone GPU mean 321 ms, and 2.49x cross-timezone CPU/GPU speedup.

AI assistance: Codex was used to analyze and prepare this change.

### Checklists

Documentation
- [ ] Updated for new or modified user-facing features or behaviors
- [x] No user-facing change

Testing
- [x] Added or modified tests to cover new code paths
- [ ] Covered by existing tests
      (Please provide the names of the existing tests in the PR description.)
- [ ] Not required

Performance
- [x] Tests ran and results are added in the PR description
- [ ] Issue filed with a link in the PR description
- [ ] Not required

### CODE REVIEW

File: tests/src/test/scala/com/nvidia/spark/rapids/timezone/OrcTimezoneSuite.scala [hunk 1]
```diff
@@ -143,6 +149,8 @@ class OrcTimezoneSuite extends SparkQueryCompareTestSuite {
     Seq(
       newYorkHistoricalTsUs,
       shanghaiHistoricalTsUs,
+      newYorkLmtTransitionTsUs,
+      portMoresbyHistoricalTsUs,
```
Comment (firestarman) [inline tests/src/test/scala/com/nvidia/spark/rapids/timezone/OrcTimezoneSuite.scala:153]: Could you also add `Math.floorDiv(portMoresbyHistoricalTsUs, TimeUnit.SECONDS.toMicros(1))` to the BIGINT schema-evolution values and a `Pacific/Port_Moresby` reader pair to `timestampSchemaEvolutionZonePairs`? This historical/LMT case is currently exercised only by the physical-timestamp matrix, so the distinct integer-to-timestamp fused conversion path remains uncovered.

**Comment by res-life** [comment]
build

**Comment by res-life** [comment]
build

**Comment by res-life** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcTimezoneUtils.scala [hunk 2]
```diff
@@ -113,22 +97,22 @@ object GpuOrcTimezoneUtils {
    * is computed in UTC. ORC Java computes base_timestamp in the *writer* timezone, so the
    * millis passed to `convertBetweenTimezones` already encode the writer TZ base offset.
    *
-   * To match ORC Java, the JNI `convertOrcTimezones` kernel first applies the writer TZ base
-   * offset and recomputes the negative nanos borrow, then applies any writer-to-reader TZ delta.
+   * To match ORC Java, the JNI `convertOrcTimestampToSpark` kernel first applies the writer TZ
+   * base offset and recomputes the negative nanos borrow, then applies any writer-to-reader TZ
+   * delta and Spark's historical java.util.TimeZone-to-java.time rebase.
```
Comment (igorpeshansky) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcTimezoneUtils.scala:102]: Note: https://github.com/NVIDIA/cudf/issues/24259 is tracking a cuDF-side fix for timestamp borrow (PR https://github.com/NVIDIA/cudf/issues/24265). Once that lands, cuDF will compute the base timestamp correctly even with `ignoreTimezoneInStripeFooter`, so the paragraph in lines 96-98 becomes stale and the borrow compensation in `cudf-spark-jni` becomes unnecessary. Both final paragraphs in this comment would then be able to go away.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcTimezoneUtils.scala [hunk 3]
```diff
@@ -193,48 +176,13 @@ object GpuOrcTimezoneUtils {
    */
   private def convertOrcTimestamp(
```
Comment (igorpeshansky) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcTimezoneUtils.scala:177]: Now that this is a one-liner, do we even need this function anymore? Can it be inlined in the two call sites?

If we do keep the function, a minor nit: the return type is the fully qualified `ai.rapids.cudf.ColumnVector`, even though `ColumnVector` is imported at this point, so the package prefix can be dropped.

File: tests/src/test/scala/com/nvidia/spark/rapids/OrcScanRetrySuite.scala [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+      RmmSpark.OomInjectionType.GPU.ordinal, 0)
+  }
+
+  private def assertRetrySucceeds(table: Table, tableSchema: StructType): Unit = {
+    injectGpuRetryOom()
+    withResource(GpuOrcScan.rebaseAndEvolveSchemaWithRetryAndClose(
+        table, tableSchema, timestampSchema, isSchemaCaseSensitive = true,
+        writerTimezone = ZoneId.of("UTC"), writerUsedProlepticGregorian = true)) { result =>
+      assertResult(1)(result.getRowCount)
+      assertResult(DType.TIMESTAMP_MICROSECONDS)(result.getColumn(0).getType)
```
Comment (igorpeshansky) [inline tests/src/test/scala/com/nvidia/spark/rapids/OrcScanRetrySuite.scala:58]: Nit: this doesn't actually test the retry (there's no way to know if the retry even happened). Should we replicate the [comment](https://github.com/NVIDIA/cudf-spark/blob/1e759de0/tests/src/test/scala/com/nvidia/spark/rapids/CsvScanRetrySuite.scala#L44-L45) from `CsvScanRetrySuite` here, i.e.:

```scala
      // We don't have any good way to verify that the retry was thrown, but we are going to trust
      // that it was.
```

Given that, should we at least make the assertion capable of catching a double-application (related to https://github.com/NVIDIA/cudf-spark/pull/16062#discussion_r4069137268, which talks about idempotency)? Given that the timestamp adjustment is the only non-idempotent operation, invoking this test with the `UTC` timezone wouldn't work — using a value assertion with a non-zero-offset timezone (e.g., `Asia/Shanghai`) should be sufficient.

> [!IMPORTANT]
> Note that `rebaseOrcDateTime` uses `ZoneId.systemDefault()` as the reader zone, which would make any value assertion environment-dependent. We would need to either pin/restore the default in `beforeEach`/`afterEach`, or derive the expected value from `ZoneId.systemDefault()`…

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcTimezoneUtils.scala [hunk 5]
```diff
@@ -51,44 +51,28 @@ object GpuOrcTimezoneUtils {
...
 
   /**
    * Rebase ORC legacy dates and timestamps considering writer and reader timezones.
    *
-   * Uses the JNI kernel `GpuTimeZoneDB.convertOrcTimezones` for both same- and cross-timezone
-   * reads. Even when the timezone rules match, the kernel must reconstruct the writer-specific
-   * ORC 2015 base before deciding whether to apply the negative nanos borrow.
+   * Uses the fused JNI kernel `GpuTimeZoneDB.convertOrcTimestampToSpark` for both same- and
+   * cross-timezone reads. Even when the timezone rules match, the kernel must reconstruct the
+   * writer-specific ORC 2015 base before deciding whether to apply the negative nanos borrow.
```
Comment (igorpeshansky) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcTimezoneUtils.scala:75]: As in https://github.com/NVIDIA/cudf-spark/pull/16062#discussion_r4086366002, the second sentence goes away once https://github.com/NVIDIA/cudf/issues/24259 gets fixed.
